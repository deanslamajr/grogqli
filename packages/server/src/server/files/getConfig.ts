import path from 'path';
import findUp from 'find-up';
import editJsonFile from 'edit-json-file';

const CONFIG_FILE_NAME = 'grogqli.json';

export interface GrogqliConfig {
  recordingsSaveDirectory: string;
  port: number;
  shouldDogFood: boolean;
}
const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: path.join(process.cwd(), 'grogqli'),
  port: 4000,
  shouldDogFood: false,
};

type GetGrogqliConfigValue = (configKey: keyof GrogqliConfig) => string;
let getGrogqliConfigValue: GetGrogqliConfigValue | undefined;

export const getConfig = async (): Promise<GetGrogqliConfigValue> => {
  // return early if the response is cached
  if (getGrogqliConfigValue) {
    return getGrogqliConfigValue;
  }

  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(
    async (directory) => {
      const hasConfig = await findUp.exists(
        path.join(directory, CONFIG_FILE_NAME)
      );
      return hasConfig ? directory : undefined;
    },
    { type: 'directory' }
  );

  let configFromFile: editJsonFile.JsonEditor;

  if (configLocation) {
    const configFileAbsolutePath = path.join(configLocation, CONFIG_FILE_NAME);
    configFromFile = editJsonFile(configFileAbsolutePath);
  }

  getGrogqliConfigValue = (configKey) => {
    if (configKey === undefined) {
      throw new Error(
        'Invalid use of getGrogqliConfigValue: must provide a key'
      );
    }
    let configValue: any;

    // First, try to resolve the value from the given config
    if (configLocation && configFromFile) {
      configValue = configFromFile.get(configKey);

      // Make 'recordingsSaveDirectory' an abs path
      if (
        configKey === 'recordingsSaveDirectory' &&
        configValue !== undefined
      ) {
        configValue = path.join(configLocation, configValue);
      }
    }

    // If config doesn't have a value, use the default
    if (configValue === undefined) {
      configValue = DEFAULT_CONFIG[configKey];
    }

    return configValue;
  };

  return getGrogqliConfigValue;
};
