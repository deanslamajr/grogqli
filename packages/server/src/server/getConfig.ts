import path from 'path';
import findUp from 'find-up';

const CONFIG_FILE_NAME = 'grogqli.json';

interface GrogqliConfig {
  recordingsSaveDirectory: string;
  recordingsFilename: string;
}

const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: '/home/dhan/codes/grogqli',
  recordingsFilename: 'groql_rec'
};
export const getConfig = async (): Promise<GrogqliConfig> => {
  console.log('process.cwd()', process.cwd());
  // TODO
  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(CONFIG_FILE_NAME);

  console.log('configLocation', configLocation)

  // TODO
  // if a config doesn't have a value defined for a particular lookup
  // use the value found in DEFAULT_CONFIG
  return DEFAULT_CONFIG;
}