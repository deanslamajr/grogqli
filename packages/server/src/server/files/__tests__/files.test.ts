import path from 'path';
import {
  getSchema,
  getWorkflowById,
  getTypeRecording,
  openTypeNameToIdMapping,
} from '..';
import { getConfig } from '../getConfig';

jest.mock('../getConfig');

describe('files', () => {
  beforeEach(() => {
    const mockedGetConfig = getConfig as jest.MockedFunction<typeof getConfig>;
    mockedGetConfig.mockImplementation(async () => () => {
      return path.join(__dirname, 'grogqli');
    });
  });

  describe('getWorkflowById', () => {
    it('should return the given workflow file contents', async () => {
      const actual = await getWorkflowById('someWorkflowId');
      expect(actual).toMatchSnapshot();
    });

    describe('if workflowId is not passed', () => {
      it('should return null', async () => {
        const actual = await getWorkflowById('');
        expect(actual).toBeNull();
      });
    });

    describe('if the file doesnt exist', () => {
      it('should return null', async () => {
        const actual = await getWorkflowById('workFlowDoesntExist');
        expect(actual).toBeNull();
      });
    });
  });

  describe('getTypeRecording', () => {
    it('should return the correct type recording', async () => {
      const typeId = 'someTypeId';
      const recordingId = 'someTypeRecordingId';
      const actual = await getTypeRecording({ typeId, recordingId });
      expect(actual).toMatchSnapshot();
    });

    describe('if file for given typeId does not exist', () => {
      it('should throw', async () => {
        const typeId = 'nonexistentTypeId';
        const recordingId = 'someTypeRecordingId';
        await expect(
          getTypeRecording({ typeId, recordingId })
        ).rejects.toThrow();
      });
    });

    describe('if type file doesnt have a recording for the given recording id', () => {
      it('should throw', async () => {
        const typeId = 'someTypeId';
        const recordingId = 'nonexistentRecordingId';
        await expect(
          getTypeRecording({ typeId, recordingId })
        ).rejects.toThrow();
      });
    });
  });

  describe('getSchema', () => {
    it('should return the schema file associated with the given schemaId', async () => {
      const schemaId = 'someSchemaId';
      const actual = await getSchema(schemaId);
      expect(actual.id).toBe(schemaId);
    });

    describe('if a schema file cannot be found that matches the given schemaId', () => {
      it('should throw', async () => {
        await expect(getSchema('nonexistentSchemaId')).rejects.toThrow();
      });
    });
  });

  describe('openTypeNameToIdMapping', () => {
    it('should return the operations file associated with the given schemaId', async () => {
      const actual = await openTypeNameToIdMapping('someSchemaId');
      expect(actual).toMatchSnapshot();
    });

    describe('if an operations file cannot be found that matches the given schemaId', () => {
      it('should throw', async () => {
        await expect(
          openTypeNameToIdMapping('nonexistentSchemaId')
        ).rejects.toThrow();
      });
    });
  });
});
