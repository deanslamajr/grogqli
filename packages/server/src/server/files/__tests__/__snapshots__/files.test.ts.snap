// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`files getOperationsData should return the operations file associated with the given schemaId 1`] = `
Object {
  "operations": Object {
    "GetChores": Object {
      "id": "llK98u0",
      "name": "GetChores",
    },
    "someOpName": Object {
      "id": "someOpId",
      "name": "someOpName",
    },
  },
  "version": 1,
}
`;

exports[`files getTypeRecording should return the correct type recording 1`] = `
Object {
  "id": "someTypeRecordingId",
  "value": Object {
    "getAccountFromSession": "someAccountRecordingId",
    "getChores": "someGetChoresResponseRecordingId",
  },
  "version": 1,
}
`;

exports[`files getWorkflowById should return the given workflow file contents 1`] = `
Object {
  "description": "some markdown text that gives context on this workflow",
  "id": "someWorkflowId",
  "name": "Some workflow",
  "operationRecordings": Object {
    "llK98u0": Object {
      "opRecordingId": "q7f66d54",
    },
    "someOpId": Object {
      "opRecordingId": "someOpRecordingId",
    },
  },
  "version": 1,
}
`;

exports[`files openTypeNameToIdMapping should return the operations file associated with the given schemaId 1`] = `
Object {
  "types": Object {
    "Chore": Object {
      "id": "bb333456",
      "name": "Chore",
    },
    "GetChoresResponse": Object {
      "id": "123j6tgc",
      "name": "GetChoresResponse",
    },
    "Query": Object {
      "id": "xhg87iT1",
      "name": "Query",
    },
    "doesntHaveATypeFile": Object {
      "id": "doesntHaveATypeFileId",
      "name": "doesntHaveATypeFile",
    },
    "someTypeName": Object {
      "id": "someTypeId",
      "name": "someTypeName",
    },
  },
  "version": 1,
}
`;
